[WIP] Add support for dynamic template options

allouis opened the pull request:
This works as is - have tested with Ghost.

I think this can be improved, the name `_templateOptions` isn't exactly friendly, though I think in order to not break things, we should give it an obscure/prefixed name, maybe even `___templateOptions`.

To keep things a nice dev experience, we might want to attach some helper methods to the `ExpressHbs` instance, so we can read, set, update etc... the dynamic options?

```js
ExpressHbs.prototype.getDynamicTemplateOptions = function (locals) {
  return locals.___templateOptions;
}

ExpressHbs.prototype.setDynamicTemplateOptions = function (locals, options) {
  locals.___templateOptions = options;
}

ExpressHbs.prototype.updateDynamicTemplateOptions = function (locals, options) {
  _.merge(locals.___templateOptions, options);
}
```

It might be worth using the word `dynamic` in the property name, but to start with I haven't as handlebars has no concept of that.

allouis committed:
Added support for express4 in test helpers createLocals

allouis committed:
Added support for dynamic template options from locals

allouis committed:
Added tests for dynamic template options

ErisDS commented:
Seeing as locals are already passed to the templates, don't these end up being passed twice?

Thinking on the other stuff 🤔

allouis commented:
Yeah that's a good point - so maybe we want to do smth like
```js
var localTemplateOptions = locals._templateOptions || {};
delete locals._templateOptions;
res = template(
  locals,
  _.merge({}, self._options.templateOptions, localTemplateOptions)
);
```

Thanks for taking a look :relaxed:

ErisDS commented:
One thing maybe worth considering `templateOptions` is slightly confusing because as far as I can tell the handlebars `template()` function only accepts `{data: }`. Although the current naming is future proof.

I think here in handlebars / express land, what we're talking about is local / not local, rather than dynamic / not dynamic.

Either `locals. _templateOptions` or `locals. ___templateOptions` is fine IMO, but the method names would use the word local, not dynamic :)

Honestly, I think one underscore is enough. We'll need to update the readme/docs a bit, and mention this in the changelog for the next release, which would be a minor. 

I think it's far more likely that this new functionality is useful to folks, than that we're gonna trample someone.

allouis committed:
Added tests for removing _templateOptions from locals

allouis committed:
Removed _templateOptions from locals

allouis committed:
Updated renderTemplate to clone before modifying

This is because we should not be modifing objects that do not belong to
us. The locals object belongs to the express application, outside of
express-hbs.

allouis committed:
Added getter/setter methods for localTemplateOptions

Used the word update here rather than set to match the global template options method

allouis committed:
Updated renderTemplate to use {get,update}LocalTemplateOptions

allouis committed:
Added getTemplateOptions method for completeness

allouis committed:
Renamed dynamic -> local

ErisDS commented on lib/hbs.js:
```diff
@@ -472,7 +484,10 @@ ExpressHbs.prototype.___express = function ___express(filename, source, options,
     try {
-      res = template(locals, self._options.templateOptions);
+      var localTemplateOptions = self.getLocalTemplateOptions(locals);
+      var localsClone = _.extend({}, locals);
```
purely out of interest - is there a reason for this syntax instead of `_.cloneDeep()` ?

allouis replied:
kiiinda - we're only cloning so that we can effectively delete the `_templateOptions` without screwing with the `locals` object from perspective of the user/function that is calling us.

So we only need to clone as much as the first level so we can get rid of `_templateOptions`

Having said that, i'm not sure how many data modifications can be made at the handlebars level - so it might be worth deep cloning the entire `locals` object, what do you think?
